syntax = "proto3";

option go_package = "protos/gen/intersection";

service OptimisationService {
  rpc RunOptimisation(OptimisationParameters) returns (OptimisationParameters);
}

message OptimisationParameters {
  OptimisationType optimisation_type = 1;
  SimulationParameters parameters = 2;
}

// Traffic light timing parameters
message SimulationParameters {
  IntersectionType intersection_type = 2;
  int32 green = 3;
  int32 yellow = 4;
  int32 red = 5;
  int32 speed = 6;
  int32 seed = 7;
}

// Enums for different types and statuses
enum IntersectionStatus {
  INTERSECTION_STATUS_UNOPTIMISED = 0;
  INTERSECTION_STATUS_OPTIMISING = 1;
  INTERSECTION_STATUS_OPTIMISED = 2;
  INTERSECTION_STATUS_FAILED = 3;
}

enum TrafficDensity {
  TRAFFIC_DENSITY_LOW = 0;
  TRAFFIC_DENSITY_MEDIUM = 1;
  TRAFFIC_DENSITY_HIGH = 2;
}

enum OptimisationType {
  OPTIMISATION_TYPE_NONE = 0;
  OPTIMISATION_TYPE_GRIDSEARCH = 1;
  OPTIMISATION_TYPE_GENETIC_EVALUATION = 2;
}

enum IntersectionType {
  INTERSECTION_TYPE_UNSPECIFIED = 0;
  INTERSECTION_TYPE_TRAFFICLIGHT = 1;
  INTERSECTION_TYPE_TJUNCTION = 2;
  INTERSECTION_TYPE_ROUNDABOUT = 3;
  INTERSECTION_TYPE_STOP_SIGN = 4;
}
